From b9d414803e77634069e6327cbc21de0c71d5978e Mon Sep 17 00:00:00 2001
From: mj <mj@maxlinear.com>
Date: Thu, 29 Jan 2026 22:33:35 -0800
Subject: [PATCH] WLANRTSYS-99574 Ignore duplicate RADAR events on non-master
 VAPs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Issue : Failed to switch second best channel due to duplicate RADAR
event received from kernel.

Root Cause: Linux cfg80211 commit c6e7f4f3b3 ("cfg80211: send radar events to all
interfaces") changed DFS event propagation by sending RADAR notifications
to all registered VAPs instead of only the interface that originally
detected radar.

Fix: Update hostapd DFS handling to ignore RADAR events received on
non‑master VAPs. Only the master VAP is allowed to process DFS/RADAR
notifications, preventing duplicate handling while maintaining correct
DFS compliance with the kernel’s updated event propagation.
---
 src/ap/drv_callbacks.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index abd73f8fa..9469a6f7d 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -2906,7 +2906,17 @@ void wpa_supplicant_event(void *ctx, enum wpa_event_type event,
 	wpa_dbg(hapd->msg_ctx, level, "Event %s (%d) received",
 		event_to_string(event), event);
 #endif /* CONFIG_NO_STDOUT_DEBUG */
-
+#ifdef CONFIG_VENDOR_MXL
+	/* Ignore DFS radar events on non‑primary BSS */
+	if (event == EVENT_DFS_RADAR_DETECTED)
+	{
+		if (hapd->iface && hapd->iface->bss[0] != hapd)
+		{
+			wpa_printf(MSG_ERROR, "nl80211: %s: Ignoring radar event on non-primary BSS", hapd->conf->iface);
+			return;
+		}
+	}
+#endif /* CONFIG_VENDOR_MXL */
 	switch (event) {
 	case EVENT_MICHAEL_MIC_FAILURE:
 		michael_mic_failure(hapd, data->michael_mic_failure.src, 1);
-- 
2.43.0

